Wiki restructuring - CI environment (#1229)
Azure/ResourceModules: docs/wiki/The CI environment - Pipeline design.md

> This section provides an overview of the design principles followed by the CARML CI environment pipelines.

Reviewer: ```suggestion
This section provides an overview of the design principles applied to the CARML CI environment pipelines.
```

  - [Wiki pipeline](#wiki-pipeline)

---

# Module Pipelines

The repository hosts one pipeline for each module in the CARML library.

The purpose of each module pipeline is twofold:

1. **Validation**: To ensure the modules hosted by the CARML library are valid and can perform the intended deployments.
1. **Publishing**: To publish _versioned_ and already validated modules to one or multiple target locations, from where they can be referenced by solutions consuming them.

As such each pipeline can be mapped to `Phases 1 and 2` described in the [Deployment flow](./The%20context%20-%20CARML%20CI%20environment#deployment-flow) section.

<img src=".\media\CIEnvironment\pipelineDesign.png" alt="Pipeline phases" height="500">

The following paragraphs provide an overview of the different phases and shared logic the module pipelines use.

- [Pipeline phases](#pipeline-phases)
- [Module pipeline inputs](#module-pipeline-inputs)

## Pipeline phases

This paragraph provides an overview of the three phases executed by each module pipeline. Further details about the implementation and design of each phase are provided in the dedicated pages linked below.

> 1. **Static Validation**: Executes a set of static Pester tests against the module template, to ensure they comply with the intended module design principles. Further details for this phase are provided by the corresponding Wiki section [Static validation](./The%20CI%20environment%20-%20Static%20validation).

Reviewer: For some reason I cannot add a suggestion and have to just add a comment:

1. **Static Validation**: Executes a set of static Pester tests against the module and its templates to ensure they comply with our design principles. Further details for this phase are provided by the corresponding Wiki section [Static validation](./The%20CI%20environment%20-%20Static%20validation).

Author: ```suggestion
1. **Static Validation**: Executes a set of static Pester tests against the module and its templates to ensure they comply with our design principles. Further details for this phase are provided by the corresponding Wiki section [Static validation](./The%20CI%20environment%20-%20Static%20validation).
```

> 1. **Deployment Validation**: An actual Azure deployment is run against a sandbox subscription leveraging a predefined set of parameter files, each validating a different configuration of the same Azure resource in parallel. The test suite is cleaned up by default removing all test resources after deployment. Further details for this phase are provided by the corresponding Wiki section [Deployment validation](./The%20CI%20environment%20-%20Deployment%20validation).

Reviewer: Same problem as before, so here we go:

> 1. **Deployment Validation**: An actual Azure deployment is run against a sandbox subscription leveraging a predefined set of parameter files, each validating a different configuration of the same Azure resource in parallel. The test suite is cleaned up by default, removing all test resources post-deployment. Further details for this phase are provided by the corresponding Wiki section [Deployment validation](./The%20CI%20environment%20-%20Deployment%20validation).

  <img src=".\media\CIEnvironment\modulePipelineInput.png" alt="Module Pipeline Input" height="300">

---

# Platform pipelines

In addition to module pipelines, the repository includes several platform pipelines covering further tasks as described below.

- [Dependencies pipeline](#dependencies-pipeline)
- [ReadMe pipeline](#readme-pipeline)
- [Wiki pipeline](#wiki-pipeline)

## Dependencies pipeline

In order to successfully run module pipelines to validate and publish CARML modules to the target environment, certain Azure resources need to be deployed beforehand.

For example any instance of the [Virtual Machine] module needs an existing virtual network to be connected to and a key vault hosting its required local admin credentials to be referenced.

The dependencies pipeline covers this requirement and is intended to be run before executing module pipelines successfully.

You can find further details about the dependencies pipeline in the  [Getting started - Dependency pipeline](./Getting%20started%20-%20Dependency%20pipeline) section.

> ### Dependencies pipeline inputs

Reviewer: Should we move this into the [Getting started - Dependency pipeline](./Getting%20started%20-%20Dependency%20pipeline) section?

Reviewer: The thought came up as I figured somebody processing the Getting-started may wonder about these inputs too

Author: I'm also in doubt on what to do with the dependency sections. We probably have too many details in the getting started which is intended to be a swift step by step guide on how to deploy dependencies. We should maybe revisit that offline a bit.

Author: Dependencies pipeline docs updates after syncing offline:
- `Getting started - Setup environment`, dependencies pipeline paragraph: contains link to pipeline usage 
- `CI environment - Pipeline usage`: contains step by step instructions on how to run the pipeline
- `CI environment - Pipeline design`: contains details on the design, e.g. deployed resources and key vault secrets and is referenced by both pages above

The `Getting started - dependency pipeline` page has been removed and its content included in the `CI environment - pipeline design` page in the `dependencies pipeline` section

The dependencies pipeline comes with the following runtime parameters:

- `'Branch' dropdown`: A dropdown to select the branch to run the pipeline from.
- `'Enable SqlMI dependency deployment' switch`: Can be enabled or disabled and controls whether the dependencies for the [SQL managed instance] module are deployed during execution. It is disabled by default.
- `'Enable deployment of a vhd stored in a blob container' switch`: Can be enabled or disabled and controls whether including the baking of a VHD and subsequent backup to a target storage blob container during the execution. This is a dependency for the [Compute Images] and [Compute Disks] modules. This task requires up to two hours completion and is disabled by default.

  <img src=".\media\CIEnvironment\dependencyPipelineInput.png" alt="Dependency Pipeline Input" height="300">

## ReadMe pipeline

> The repository includes two major ReadMe files that require to stay in sync with the provided modules.

Reviewer: ```suggestion
The repository includes two major ReadMe files that should stay in sync with the available modules.
```

The first can be found in the repository root (`README.md`) and the second in the modules folder (`arm/README.md`).

The ReadMe pipeline is triggered each time changes are pushed to the `main` branch and only if a template in the `arm` folder is being altered.

Once triggered, the pipeline crawls through the library and updates the tables in each corresponding ReadMe file, creating links to the corresponding pipeline runs and updating the list of entries.

## Wiki pipeline

> The purpose of the Wiki pipeline is to sync any files from the `docs/wiki` folder to the GitHub wiki repository. It is triggered each time changes are pushed to the `main` branch and only if files in the `docs/wiki` folder are altered.

Reviewer: Also available for ADO: 

```suggestion
The purpose of the Wiki pipeline is to sync any files from the `docs/wiki` folder to the wiki repository. It is triggered each time changes are pushed to the `main` branch and only if files in the `docs/wiki` folder are altered.
```
